hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapReplicationStateHolder.java

```diff
@@ -202,16 +199,40 @@ private void addIndexes(String mapName, Collection<IndexInfo> indexInfos) {
                 }
             }
         } else {
-            Indexes indexes = mapContainer.getIndexes(mapReplicationOperation.getPartitionId());
+            Indexes indexes = mapContainer.getIndexes(operation.getPartitionId());
             for (IndexInfo indexInfo : indexInfos) {
                 indexes.addOrGetIndex(indexInfo.getAttributeName(), indexInfo.isOrdered());
             }
         }
     }
 
+    private static SerializationService getSerializationService(RecordStore recordStore) {
+        return recordStore.getMapContainer().getMapServiceContext().getNodeEngine().getSerializationService();
+    }
+
     @Override
+    @SuppressWarnings("checkstyle:npathcomplexity")
     public void writeData(ObjectDataOutput out) throws IOException {
-        out.writeInt(data.size());
+        out.writeInt(storesByMapName.size());
+
+        for (Map.Entry<String, RecordStore<Record>> entry : storesByMapName.entrySet()) {
+            String mapName = entry.getKey();
+            RecordStore recordStore = entry.getValue();
+
+            SerializationService ss = getSerializationService(recordStore);
+
+            out.writeUTF(mapName);
+            out.writeInt(recordStore.size());
+
+            Iterator<Record> iterator = recordStore.iterator();
+            while (iterator.hasNext()) {
+                Record record = iterator.next();
+                RecordReplicationInfo replicationInfo = operation.toReplicationInfo(record, ss);
+                out.writeObject(replicationInfo);
+            }
+        }
+
+
         for (Map.Entry<String, Set<RecordReplicationInfo>> dataEntry : data.entrySet()) {
```

Comment:
The `data` map shouldn't be used at all in the write path, this block is now redundant.

Reply:
good catch, removed that part